source/common/router/router_ratelimit.cc

```diff
@@ -109,27 +107,40 @@ void RateLimitPolicyEntryImpl::populateDescriptors(
 
 RateLimitPolicyImpl::RateLimitPolicyImpl(const Json::Object& config) {
   if (config.hasObject("rate_limits")) {
-    std::vector<std::unique_ptr<RateLimitPolicyEntry>> rate_limit_policy;
-    std::vector<std::reference_wrapper<const RateLimitPolicyEntry>> rate_limit_policy_reference;
     for (const Json::ObjectPtr& rate_limit : config.getObjectArray("rate_limits")) {
       std::unique_ptr<RateLimitPolicyEntry> rate_limit_policy_entry(
           new RateLimitPolicyEntryImpl(*rate_limit));
-      rate_limit_policy_reference.emplace_back(*rate_limit_policy_entry);
-      rate_limit_policy.emplace_back(std::move(rate_limit_policy_entry));
+
+      // Every rate limit policy is applicable to the default stage.
+      default_rate_limit_entries_reference_.emplace_back(*rate_limit_policy_entry);
+
+      // A non-zero stage value indicates a rate limit policy that applies to a non-default stage in
+      // addition to the default stage.
+      int64_t stage = rate_limit_policy_entry->stage();
+      if (stage != 0) {
+        rate_limit_entries_reference_[stage].emplace_back(*rate_limit_policy_entry);
```

Comment:
In practice, we are going to have very few stages. Can we use a vector here so that we can do a direct bounds check and index and avoid a map lookup.